libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/dialogs/QnAMakerDialog.java

```diff
@@ -905,27 +910,29 @@ private static void resetOptions(DialogContext dc, QnAMakerDialogOptions dialogO
                 result.add(response.getAnswers()[0]);
             }
 
-            stepContext.setValue(ValueProperty.QNA_DATA, result);
+            stepContext.getValues().put(ValueProperty.QNA_DATA, result);
```

Comment:
Be careful that here you are changing the behavior.

The line you are removing says: I'll assign a new value using the object `stepContext`
The line you are adding says: I'll get the values apart from the object of `stepContext` and I'll assign a new value.

I mean, the change won't be reflected in the `stepContext` in the last scenario

Reply:
Check these kind of changes

Reply:
In [C#](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs#L675) what this does is: call the Values method which does a Get of the Values `(IDictionary<string, object>)` variable of the [WaterfallStepContext](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs/WaterfallStepContext.cs#L94). In java is the same, call the `getValues()` method which return `this.values`, `values` is a `Map<String, Object>` and then it putting an element. I think the implementation is the same in C# and Java

Reply:
Ack, we will review during the execution of unit tests

Reply:
I agree up to that `this.values`, but its important to know if `this.values` is by value or by reference. If it is by value, it will return a copy of the map, not the same map. Even if you modify what it returns, it won't be reflected on the original map.

If that is the case, you need to retrieve the map, modify it, and set it again with `setValues`, which is why it has `getValues` and `setValues` methods.

If its by reference, no problem then.

Reply:
Refer to [this post](https://stackoverflow.com/questions/7297937/does-java-return-by-reference-or-by-value)

Reply:
Yes, you're right @VictorGrycuk, I can do what you say, or wait until test this with the unit tests